include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+    va_region(va_region &&) = default;
+    va_region &operator=(va_region &&) = default;
+
+    ~va_region() {
+      purge();
+    }
+
+    int num_blocks() const {
+      return mapping.size();
+    }
+
+    void purge() {
+      CUdeviceptr ptr = address_range.ptr();
+      for (size_t i = 0; i < mapping.size(); i++, ptr += block_size) {
+        auto &h = mapping[i];
+        if (h) {
+          cuvm::Unmap(ptr, block_size);
+          CUDA_DTOR_CALL(cuMemRelease(h));
+          h = {};
+        }
+      }
+      mapped.fill(false);
+      available.fill(false);
+    }
+
+    CUdeviceptr block_dptr(ptrdiff_t block_index) const noexcept {
+      return address_range.ptr() + block_index * block_size;
+    }
+
+    template <typename T = void>
+    T *block_ptr(ptrdiff_t block_index) const noexcept {
+      return reinterpret_cast<T*>(block_dptr(block_index));
+    }
+
+    void map_block(int block_idx, cuvm::CUMem mem) {
+      assert(mem.size() == block_size);
+      cuvm::Map(block_dptr(block_idx), mem);
+      mapping[block_idx] = mem.release().first;
+      mapped[block_idx] = true;
+      available[block_idx] = true;
+      available_blocks++;
+    }
+
+    cuvm::CUMem unmap_block(ptrdiff_t block_idx) {
+      assert(available[block_idx]);
+      cuvm::Unmap(block_dptr(block_idx), block_size);
+      cuvm::CUMem mem({mapping[block_idx], block_size });
+      mapping[block_idx] = 0;
+      mapped[block_idx] = false;
+      available[block_idx] = false;
+      available_blocks--;
+      return mem;
+    }
+
+    void append(va_region &&other) {
+      assert(address_range.end() == other.address_range.ptr());
+      assert(block_size == other.block_size);
+      address_range.size() += other.address_range.size();
+      int other_blocks = other.num_blocks();
+      int old_blocks = num_blocks();
+      mapping.resize(old_blocks + other_blocks);
+      for (int i = 0; i < other_blocks; i++) {
+        mapping[old_blocks + i] = other.mapping[i];
+        other.mapping[i] = {};
+      }
+      other.mapping = {};
+
+      mapped.append(other.mapped);
+      available.append(other.available);
+
+      other.available = {};
+      other.mapped = {};
+      available_blocks += other.available_blocks;
+      other.available_blocks = 0;
+      assert(mapping.size() == mapped.size());
+      assert(mapping.size() == available.size());
+    }
+
+    void resize(int new_num_blocks) {
+      if (new_num_blocks < num_blocks()) {
+        int no_longer_in_range = 0;
+        for (int i = new_num_blocks; i < num_blocks(); i++) {
+          if (mapping[i]) {
+            cuvm::Unmap(block_ptr(i), block_size);
+            CUDA_DTOR_CALL(cuMemRelease(mapping[i]));
+            mapping[i] = {};
+            if (available[i])
+              no_longer_in_range++;
+          }
+        }
+        available_blocks -= no_longer_in_range;
+      }
+      mapping.resize(new_num_blocks);
+      mapped.resize(new_num_blocks);
+      available.resize(new_num_blocks);
```

Comment:
```suggestion
      mapped.resize(new_num_blocks, false);
      available.resize(new_num_blocks, false);
```

Reply:
That's the default. When you resize a vector of primitive type, it's zero-filled - it's the same with bitmask.

Reply:
ok